halo2_proofs/src/poly/commitment/msm.rs

```diff
@@ -3,15 +3,19 @@ use crate::arithmetic::{best_multiexp, CurveAffine};
 use ff::Field;
 use group::Group;
 
+use std::collections::BTreeMap;
+
 /// A multiscalar multiplication in the polynomial commitment scheme
 #[derive(Debug, Clone)]
 pub struct MSM<'a, C: CurveAffine> {
     pub(crate) params: &'a Params<C>,
     g_scalars: Option<Vec<C::Scalar>>,
     w_scalar: Option<C::Scalar>,
     u_scalar: Option<C::Scalar>,
-    other_scalars: Vec<C::Scalar>,
-    other_bases: Vec<C>,
+    // TODO: we could make C: Ord somehow and use it as the key instead of this
+    // hacky approach that wastes ~64 bytes of memory for the pasta curves per
+    // entry.
+    other: BTreeMap<(C::Base, C::Base), (C::Scalar, C)>,
```

Comment:
There's a wonderfully awful trick we could use: it's sufficient for the map to only include points with a "positive" y-coordinate (i.e. ỹ = 1). To add an entry for a point with ỹ = 0, subtract the scalar instead of adding it (which is equivalent to adding it with the negated point).

Reply:
You could save another 32 bytes by using BTreeMap<C::Base, (C::Scalar, C::Base)> where the key is the x-coordinate and the second element of the tuple is the y-coordinate.

Reply:
I'm going to try implementing this as a PR onto this one.

Reply:
https://github.com/ebfull/halo2/pull/1 (works, but I want to clean it up a bit and expand the test).